### DISCUSSION TITLE
globalThis is not defined with RSC

### DISCUSSION THREAD

**Discussion by Moshyfawn** [vercel/next.js#49771] [Help]
### Summary

I'm trying to run a new Next.js 13 app in an older environment and I keep getting a `globalThis is not defined` compile error _(full error is in the "Additional information" section below)_.

The environment I'm working in is **Android 7** _(API 25)_ **WebView** _(**Chrome 55.0**)_, so I was getting `[Uncaught SyntaxError: Unexpected token ...]` initially. I've provided a custom `.babelrc` config with a `@babel/plugin-proposal-object-rest-spread` polyfill. But now I'm getting `globalThis is not defined` and it doesn't seem like I can use a polyfill to fix it.

Here's my .babelrc for reference:
```json
{
  "presets": [
    "next/babel"
  ],
  "plugins": [
    "@babel/plugin-proposal-object-rest-spread"
  ]
}

```

### Additional information

```js
I/chromium: [INFO:CONSOLE(36590)] "%cDownload the React DevTools for a better development experience: https://reactjs.org/link/react-devtools", source: webpack-internal:///(app-client)/./node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (36590)
I/chromium: [INFO:CONSOLE(56)] "Warning: An error occurred during hydration. The server HTML was replaced with client content in <%s>.", source: webpack-internal:///(app-client)/./node_modules/next/dist/client/app-index.js (56)
I/chromium: [INFO:CONSOLE(1058)] "Uncaught ReferenceError: globalThis is not defined", source: webpack-internal:///(app-client)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-client.browser.development.js (1058)
I/chromium: [INFO:CONSOLE(56)] "The above error occurred in the <ServerRoot> component:
    
        at ServerRoot (webpack-internal:///(app-client)/./node_modules/next/dist/client/app-index.js:172:25)
        at RSCComponent
        at Root (webpack-internal:///(app-client)/./node_modules/next/dist/client/app-index.js:196:25)
    
    Consider adding an error boundary to your tree to customize error handling behavior.
    Visit https://reactjs.org/link/error-boundaries to learn more about error boundaries.", source: webpack-internal:///(app-client)/./node_modules/next/dist/client/app-index.js (56)
I/chromium: [INFO:CONSOLE(56)] "Error: There was an error while hydrating. Because the error happened outside of a Suspense boundary, the entire root will switch to client rendering.", source: webpack-internal:///(app-client)/./node_modules/next/dist/client/app-index.js (56)
I/chromium: [INFO:CONSOLE(25067)] "Uncaught ReferenceError: globalThis is not defined", source: webpack-internal:///(app-client)/./node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (25067)

Operating System:
  Platform: darwin
  Arch: arm64
Binaries:
  Node: 20.0.0
  npm: 9.6.4
  Yarn: 1.22.19
  pnpm: 7.29.1
Relevant packages:
  next: 13.4.2
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.0.4
```

**Comment by icyJoseph** [comment]
So, you are not using `globalThis` directly, right? The problem is that the environment you are using the run the app doesn't support `globalThis` correct?

The issue is then, how to polyfill globalThis (and other potentially missing API's) in your environment. You should probably read through: https://nextjs.org/docs/architecture/supported-browsers#custom-polyfills, but I am not sure how to do this best for the App Router, and also, not sure there's a good globalThis polyfill out there, except for maybe: https://mathiasbynens.be/notes/globalthis, which I guess you could load through a beforeInteractive next/script, pointing to:

```jsx
// pseudo-code

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <body className={inter.className}>{children}</body>
      <Script
        strategy="beforeInteractive"
        src="https://unpkg.com/@ungap/global-this@0.4.4/min.js"
      ></Script>
    </html>
  );
}

```

If you see a `lint` warning about `_document`, I think you can ignore it.

**Reply by emmgfx** [reply]
@icyJoseph with this, we'll be loading the polyfill for everyone in client, outside the build time?

There isn't a official answer? I'm having the same issue, and also not using globalthis directly,

**Reply by bbuergel** [reply]
Having the same issue with `globalThis` not working for polyfills in a Webview 33 environment.
Whenever I try to add a `<Script>` with `strategy="beforeInteractive"` in the root layout, I get React hydration errors saying the Initial UI does not match what was rendered on the server.
Running into a wall here not being able to make any traction on polyfills. What's the best approach to add custom polyfills with the App router?

**Reply by eijil** [reply]
Write the script tag into the Body, there will be no error warnings

**Comment by ben-Run** [comment]
@icyJoseph
Many thanks

**Comment by one-pieces** [comment]
@icyJoseph I got the same error in the older webview, I think we can add the `globalThis` polyfill in  https://github.com/vercel/next.js/blob/canary/packages/next-polyfill-nomodule/src/index.js

**Comment by Rieranthony** [comment]
I'm having the same issue, my sentry is bombarded by this error, any response from the NextJS team maybe? Thanks! 🙏🏻

**Reply by Kyoss79** [reply]
I did it the following way. (Just for Info: NextJS 14 - App Directory)

In layout.tsx inside <head> I added the following:

`<script src="https://unpkg.com/@ungap/global-this@0.4.4/min.js" noModule async></script>`

It's included in the head just as intended.

**Comment by zchwyng** [comment]
Also seeing the same issue.

**Comment by 2020191543223** [comment]
I Also seeing the same issue. The environment I'm working in is Android 9

**Comment by alaa-m1** [comment]
I've encountered the same error even with newly created projects, and I tested it using Opera 57
I hope the Next.JS team will take this into consideration in future releases
